### DISCUSSION TITLE
Lesson 7: Incorrect Decimals? 06:29:31

### DISCUSSION THREAD

**Discussion by abdulrabbani00** [smartcontractkit/full-blockchain-solidity-course-py#498] [Q&A] [answered]
For the first test via brownie, I believe the decimals being used are incorrect. When I run the test, I get the following:
```

    def test_get_entrance_fee():
        account = accounts[0]
        lottery = Lottery.deploy(config["networks"][network.show_active()]["eth_usd_price_feed"],
                                 {"from": account})
        assert lottery.getEntranceFee() > Web3.toWei(0.010, "ether")
>       assert lottery.getEntranceFee() < Web3.toWei(0.014, "ether")
E       AssertionError: assert 1301113216203868827 < 14000000000000000
E        +  where 1301113216203868827 = <ContractCall 'getEntranceFee()'>()
E        +    where <ContractCall 'getEntranceFee()'> = <Lottery Contract '0xE7eD6747FaC5360f88a2EFC03E00d25789F69291'>.getEntranceFee
E        +  and   14000000000000000 = <function to_wei at 0x7f7ebefc7320>(0.014, 'ether')
E        +    where <function to_wei at 0x7f7ebefc7320> = Web3.toWei

tests/test_lottery.py:11: AssertionError
========================= short test summary info ==========================
FAILED tests/test_lottery.py::test_get_entrance_fee - AssertionError: ass...
============================ 1 failed in 6.33s =============================
Terminating local RPC client...
```

Taking a look at the length of both sides. I get the following:
Length of the return from `lottery.getEntranceFee()`: `19`
Length of the return from `Web3.toWei(0.010, "ether")`: `17`

Can someone please help me understand why we are seeing these different values?

**Comment by cromewar** [comment]
Hello @abdulrabbani00 the assertion is failing because the way you are setting up the  `Web3.toWei(0.014, "ether")` you should use just 0.14 rather thn 0.014, that additional zero is what causing your code to fail, as you can see the lenghs are mismatch, also make sure your don't have an additional zero on `getEntranceFee` function as it's length should be 18.

**Comment by abdulrabbani00** [comment]
@cromewar - Good catch on the `Web3.toWei(0.014, "ether")`. I am now getting a length of 18 from Web3. But I am still getting 19 decimal places from the `getEntranceFee`. Here is the code (which looks to be the same as the code from the repo):
```Solidity
    constructor(address _priceFeedAddress) public {
        usdEntryFee = 50 * (10**18);
        ethUsdPriceFeed = AggregatorV3Interface(_priceFeedAddress);
        lottery_state = LOTTERY_STATE.CLOSED;
    }
```

```Solidity
    function getEntranceFee() public view returns(uint256) {
        (,int price,,,) = ethUsdPriceFeed.latestRoundData();
        uint256 adjustedprice =  uint256(price) * 10**10;//18 decimal
        // $50, $4000 / ETH
        // 50/2000
        // 50 * 10000 / 2000
        uint256 costToEnter = (usdEntryFee * 10 ** 18) / adjustedprice ;
        return costToEnter;
    }
```

**Comment by cromewar** [comment, marked as answer]
Actually my bad I was expecting another zero, but as your code does not have it, the 0.014 might be correct so the error is not by that side, the eth price for 50$ is 0.012, so your test would pass setting them bellow and above that number, however the entrance fee is returning a bigger number as it's not converting the floating numbers. Anyways you will change that test later on the course as is just intended to be used on local networks,

I have a workaround for you, don't use to Wei for this  test, use just numbers (your know 11 * 10.... and 14*10...). let us know it this works for you. 

This is how the function will like later: 
```Python

def test_get_entrance_fee():
    if network.show_active() not in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        pytest.skip()
    lottery = deploy_lottery()
    # Act
    expected_entrace_fee = Web3.toWei(0.025, "ether")
    entrance_fee = lottery.getEntranceFee()
    # Assert
    print(expected_entrace_fee)
    print(entrance_fee)
    assert expected_entrace_fee == entrance_fee
```

**Reply by abdulrabbani00** [reply]
@cromewar - thank you for the workaround. I also played around with a few variables and got the test to work, but that does not explain the behavior. Is there a reason that the value from `getEntranceFee` is incorrect? I will continue with the lesson, but this just raised a few eyebrows, especially since I don't fully understand the conversations and decimal places logic just yet.

**Reply by cromewar** [reply]
Hi @abdulrabbani00 is not that getEntranceFee is incorrect, it means is returning a value with 'x' amount of decimals which is ok, the reason we manipulate it to be 18 decimals is because is the standard for most of ERC20 tokens including Ethereum. some tokens or answers like this returns smaller numbers as its cheaper for gas cost. 

Your will struggle with this things a lot, yo can refer to this link to understand deeply how it works: https://forum.openzeppelin.com/t/understanding-decimals-for-erc20-token-creation/1821

As your problem is solved now I'll close this thread now, please feel free to make any other questions I'll do my best to answer.

best regards, Cromewar.
